services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/X509CertificateCredential.java

```diff
@@ -198,4 +261,68 @@ public final X509CertificateCredential setSecrets(final List<X509CertificateSecr
         this.secrets.addAll(secrets);
         return this;
     }
+
+    /**
+     * Applies the <em>auth-id-template</em> from the tenant's trust anchor to the client certificate's subject DN.
+     * <p>
+     * It is only applicable, if a template is configured.
+     *
+     * @param tenant The tenant information.
+     * @return the credential with the generated authentication identifier.
+     * @throws NullPointerException if the tenant is {@code null}.
+     */
+    @JsonIgnore
+    public final X509CertificateCredential applyAuthIdTemplate(final Tenant tenant) {
+        Objects.requireNonNull(tenant, "tenant information must not be null");
+
+        final String generatedAuthId = Optional.ofNullable(issuerDN)
+                .flatMap(tenant::getAuthIdTemplate)
+                .map(IdentityTemplate::new)
+                .map(t -> t.apply(getAuthId()))
+                .orElse(null);
+        final X509CertificateCredential credential = new X509CertificateCredential(getAuthId(),
+                generatedAuthId, secrets) {
+
+            @Override
+            @JsonGetter(value = RegistryManagementConstants.FIELD_GENERATED_AUTH_ID)
+            protected String getGeneratedAuthId() {
+                return super.getGeneratedAuthId();
+            }
+        };
+
+        copyCommonAttributesTo(credential);
+        return credential;
+    }
```

Comment:
I guess that this works but, pardon my French, it looks like a hack. Wouldn't it be easier to understand and also much cleaner from a design perspective to introduce two subclasses instead, one for handling credentials with a generated ID and the other for a non-generated ID?

Reply:
I noticed that anonymous class approach is employed in several places in Hono. I chose that approach than my initial one. Good that I still have a backup of `X509CertificateCredentialWithGeneratedAuthId` ;-)